examples/solvers/pdhg_denoising_ROF_algorithm_comparison.py

```diff
@@ -0,0 +1,221 @@
+
+    def __call__(self, x):
+        # Fill in proper calls to functionals here
+        self.iteration_count += 1
+        k = self.iteration_count
+
+        self.iteration_counts.append(self.iteration_count)
+        self.ergodic_iterate = (k-1)/k * self.ergodic_iterate + 1/k * x
+        self.obj_function_values.append(obj(x))
+        self.obj_function_values_ergodic.append(obj(self.ergodic_iterate))
+
+    def reset(self):
+        self.iteration_count = 0
+        self.iteration_counts = []
+        self.ergodic_iterate = 0
+        self.obj_function_values = []
+        self.obj_function_values_ergodic = []
+
+
+callback = (odl.solvers.CallbackPrintIteration() & CallbackStore())
+
+# number of iterations
+niter = 500
+
+# %% run algorithms
+
+# --- algorithm 1 --- #
+
+# Operator assignment
+op = odl.BroadcastOperator(odl.IdentityOperator(space), gradient)
+
+# Make separable sum of functionals, order must correspond to the operator K
+f = odl.solvers.SeparableSum(l2_norm, l1_norm)
+
+# Non-negativity constraint
+g = char_fun
+
+# Estimated operator norm, add 10 percent to ensure ||K||_2^2 * sigma * tau < 1
+op_norm = 1.1 * odl.power_method_opnorm(op, xstart=noisy)
+
+tau = 1.0 / op_norm  # Step size for the primal variable
+sigma = 1.0 / op_norm  # Step size for the dual variable
+
+# Starting point
+x_start = op.domain.zero()
+
+# Run algorithm 1
+x_alg1 = x_start.copy()
+callback.reset()
+callback.callbacks[1].reset()
+odl.solvers.pdhg(x_alg1, f, g, op, tau=tau, sigma=sigma, niter=niter,
+                 callback=callback)
+obj_alg1 = callback.callbacks[1].obj_function_values
+obj_ergodic_alg1 = callback.callbacks[1].obj_function_values_ergodic
+
+# --- algorithm 2 and 3 --- #
+
+# Operator assignment
+op = gradient
+
+# Assign functional f
+f = l1_norm
+
+# Create new functional that combines data fit and characteritic function
+g = odl.solvers.FunctionalQuadraticPerturb(char_fun, factr, -2 * factr * noisy)
+
+# The operator norm of the gradient with forward differences is well-known
+op_norm = np.sqrt(8) + 1e-4
+
+tau = 1.0 / op_norm  # Step size for the primal variable
+sigma = 1.0 / op_norm  # Step size for the dual variable
+
+# Run algorithms 2 and 3
+x_alg2 = x_start.copy()
+callback.reset()
+callback.callbacks[1].reset()
+odl.solvers.pdhg(x_alg2, f, g, op, tau=tau, sigma=sigma, niter=niter,
+                 gamma_primal=0, callback=callback)
+obj_alg2 = callback.callbacks[1].obj_function_values
+obj_ergodic_alg2 = callback.callbacks[1].obj_function_values_ergodic
+
+x_alg3 = x_start.copy()
+callback.reset()
+callback.callbacks[1].reset()
+odl.solvers.pdhg(x_alg3, f, g, op, tau=tau, sigma=sigma, niter=niter,
+                 gamma_primal=strong_convexity, callback=callback)
+obj_alg3 = callback.callbacks[1].obj_function_values
+obj_ergodic_alg3 = callback.callbacks[1].obj_function_values_ergodic
+
+# %% Display results
+# show images
+plt.figure(0)
+ax1 = plt.subplot(231)
+ax1.imshow(orig, clim=[0, 1])
```

Comment:
Usually one needs a conversion from ODL's `'xy'` axes to matplotlib's `'ij'` axes. Probably the images end up correctly oriented like this, but if you do, e.g., `orig.show()` the image will be rotated by 90 degrees.

When wrapping a raw array representing an image in `'ij'` convention, the conversion is `np.rot90(img, -1)`, and the other way around it is `np.rot90(odl_elem, 1)`.

It's up to you if you consider that useful here.

Reply:
I would like to use img.show() but for this example, it is most important to create plot that are next to each other. Of course the user can drag them there but I really would not recommend that to anyone. As far as I know img.show() will always open a new figure, so it cannot be used here. Any ideas how to combine these two ideas?

Reply:
Well you have a point there, and no, I don't have any idea on how to solve it. I guess this way is fine.